### PR TITLE
React-query: useTypeImports

### LINKED ISSUE #7661: typescript-react-query can generate redundant import statements
### Issue workflow progress

_Progress of the issue based on the [Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [ ] 1. The issue provides a reproduction available on [Github](https://github.com/dotansimha/graphql-code-generator-issue-sandbox-template), [Stackblitz](https://stackblitz.com/github/dotansimha/graphql-code-generator-issue-sandbox-template) or [CodeSandbox](https://codesandbox.io/s/github/dotansimha/graphql-code-generator-issue-sandbox-template)

_Make sure to fork this template and run `yarn generate` in the terminal._

_Please make sure the Codegen and plugins version under `package.json` matches yours._

- [ ] 2. A failing test has been provided

- [x] 3. A local solution has been provided

- [x] 4. A pull request is pending review

---

**Describe the bug**
`useTypeImports: true` resulted a duplicate import:
```diff
+ import type { GraphQLClient } from 'graphql-request';
+ import type { RequestInit } from 'graphql-request/dist/types.dom';
- import { GraphQLClient } from 'graphql-request';
```

**To Reproduce**
caused when config has the following:
1. `typescript-react-query` plugin
2. `graphql-request` as the fetcher
3. `useTypeImports: true`

1. My GraphQL schema:

```graphql
# any
```

4. My GraphQL operations:

```graphql
# any
```

5. My `codegen.yml` config file:

```yml
schema:
  - ${APIURL}
documents: "my-codegen-src/*.graphql"
generates:
  my-codegen-out/reactQuery.ts:
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-query
      useTypeImports: true
```

**Expected behavior**
the generated code should not attempt to import the same variable twice

**Additional context**
 - infinite query has code to handle `useTypeImport`
 https://github.com/dotansimha/graphql-code-generator/blob/ed59f81ff725a38085279f147d1f890af039867e/packages/plugins/typescript/react-query/src/fetcher-graphql-request.ts#L32-L33
 - query has code to handle `useTypeImport`
 https://github.com/dotansimha/graphql-code-generator/blob/ed59f81ff725a38085279f147d1f890af039867e/packages/plugins/typescript/react-query/src/fetcher-graphql-request.ts#L69-L70
 - mutation missed the code 
 https://github.com/dotansimha/graphql-code-generator/blob/ed59f81ff725a38085279f147d1f890af039867e/packages/plugins/typescript/react-query/src/fetcher-graphql-request.ts#L103

Sending a PR to fix.

### LINKED ISSUE #7666: typescript-react-query doesn't respect `useTypeImports` from `react-query`
### Issue workflow progress

- [ ] 1. The issue provides a reproduction available on [Github](https://github.com/dotansimha/graphql-code-generator-issue-sandbox-template), [Stackblitz](https://stackblitz.com/github/dotansimha/graphql-code-generator-issue-sandbox-template) or [CodeSandbox](https://codesandbox.io/s/github/dotansimha/graphql-code-generator-issue-sandbox-template)

- [ ] 2. A failing test has been provided

- [x] 3. A local solution has been provided

- [x] 4. A pull request is pending review

---

**Describe the bug**
when using `useTypeImports`, generated code is incorrect.  it should be:

```diff
- import { useQuery, UseQueryOptions, useMutation, UseMutationOptions } from 'react-query';
+ import { useQuery, useMutation } from 'react-query';
+ import type { UseQueryOptions, UseMutationOptions } from 'react-query';
```

**To Reproduce**
caused when config has the following:

- typescript-react-query plugin
- useTypeImports: true

1. My GraphQL schema:

```graphql
# any
```

2. My GraphQL operations:

```graphql
# any
```

3. My `codegen.yml` config file:

```yml
schema:
  - ${APIURL}
documents: "my-codegen-src/*.graphql"
generates:
  my-codegen-out/reactQuery.ts:
    plugins:
      - typescript
      - typescript-operations
      - typescript-react-query
      useTypeImports: true
```

**Expected behavior**
Types need to be tracked separately and imports should respect the `useTypeImports` flag.

**Environment:**
-any

**Additional context**
Similar to #7661 (but not identical). Sending PR to fix this.

### PR DESCRIPTION
**Pull request by jonathanstanley**
## Description

corrects behavior so that `typescript-react-query` respects `useTypeImports` flag

Related # (issue)
fixes #7666 main issue
fixes #7661 the graphql-request fetcher portion

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

tests all pass.  added test to [prove functionality](https://github.com/dotansimha/graphql-code-generator/blob/6cdb68b9123dccc80009b2af9ac66bb7de961cd9/packages/plugins/typescript/react-query/tests/react-query.spec.ts#L528).

**Test Environment**:
n/a

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by jonathanstanley** [comment]
tests pass locally.

```
yarn test packages/plugins/typescript/react-query  

Test Suites: 1 passed, 1 total
Tests:       31 passed, 31 total
Snapshots:   12 passed, 12 total
Time:        0.658 s, estimated 1 s
```
i don't know why dev-tests failing here.

**Review by charlypoly — CHANGES_REQUESTED** [review]
Hi @jonathanstanley,

Thank you for your work!
I have 2 feedback:
- It would be preferable that imports from the same source are grouped (cf inlined comment)
- Could you add tests when `useTypeImports` is used?

Thank you!

File: packages/plugins/typescript/react-query/src/visitor.ts [hunk 1]
```diff
@@ -112,10 +113,16 @@ export class ReactQueryVisitor extends ClientSideBaseVisitor<ReactQueryRawPlugin
...
     }
 
-    return [...baseImports, `import { ${Array.from(this.reactQueryIdentifiersInUse).join(', ')} } from 'react-query';`];
+    const typeImport = this.config.useTypeImports ? 'import type' : 'import';
+
+    return [
+      ...baseImports,
+      `import { ${Array.from(this.reactQueryHookIdentifiersInUse).join(', ')} } from 'react-query';`,
+      `${typeImport} { ${Array.from(this.reactQueryOptionsIdentifiersInUse).join(', ')} } from 'react-query';`,
+    ];
```
[the remark this answers is not in the document]
Comment (jonathanstanley) [reply packages/plugins/typescript/react-query/src/visitor.ts:125]: yes; i'll add a commit shortly

[the remark this answers is not in the document]
Comment (jonathanstanley) [reply packages/plugins/typescript/react-query/src/visitor.ts:125]: > * Could you add tests when `useTypeImports` is used?
it is in there... just a bit buried:
https://github.com/dotansimha/graphql-code-generator/blob/6cdb68b9123dccc80009b2af9ac66bb7de961cd9/packages/plugins/typescript/react-query/tests/react-query.spec.ts#L528

[the remark this answers is not in the document]
Comment (jonathanstanley) [reply packages/plugins/typescript/react-query/src/visitor.ts]: done:
https://github.com/dotansimha/graphql-code-generator/pull/7667/commits/799405d26a65c329a2614c5e9da1825d99c748c2#diff-38f7b7f89098e5486e40c54e00569d09eb7b873bc8d4a7118aa39a1ce1c1be68

also adjusted tests

**Review by charlypoly — APPROVED** [review]
Thank you for the fast changes @jonathanstanley! ⚡

**Pull request merged**